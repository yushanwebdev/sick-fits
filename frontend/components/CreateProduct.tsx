import useForm from '../lib/useForm';
import Form from './styles/Form';

export default function CreateProduct() {
  const { inputs, handleChange } = useForm<{
    image: string;
    name: string;
    price: number;
    description: string;
  }>({
    image: '',
    name: 'Nice Shoes',
    price: 34234,
    description: 'These are the best shoes!',
  });

  return (
    <Form
      onSubmit={(e) => {
        e.preventDefault();
        console.log('inputs', inputs);
      }}
    >
      <fieldset aria-busy>
        <label>
          Image
          <input
            required
            type="file"
            id="image"
            name="image"
            onChange={handleChange}
          />
        </label>
        <label>
          Name
          <input
            type="text"
            id="name"
            name="name"
            placeholder="Name"
            value={inputs.name}
            onChange={handleChange}
          />
        </label>
        <label>
          Price
          <input
            type="number"
            id="price"
            name="price"
            placeholder="Price"
            value={inputs.price}
            onChange={handleChange}
          />
        </label>
        <label>
          Description
          <textarea
            id="description"
            name="description"
            placeholder="Description"
            value={inputs.description}
            onChange={handleChange}
          />
        </label>
        <button type="submit">+ Add Product</button>
      </fieldset>
    </Form>
  );
}
