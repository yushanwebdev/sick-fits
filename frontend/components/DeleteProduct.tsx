import { gql, useMutation } from '@apollo/client';

const DELETE_PRODUCT_MUTATION = gql`
  mutation DeleteProduct($id: ID!) {
    deleteProduct(id: $id) {
      name
    }
  }
`;

export default function DeleteProduct({
  id,
  children,
}: {
  id: string;
  children: React.ReactNode;
}) {
  const [deleteProduct, { loading }] = useMutation(DELETE_PRODUCT_MUTATION, {
    variables: { id },
  });

  return (
    <button
      type="button"
      onClick={() => {
        // eslint-disable-next-line no-restricted-globals
        if (confirm('Are you sure you want to delete this item?')) {
          // go ahead and delete it
          deleteProduct().catch((err) => alert((err as Error).message));
        }
      }}
    >
      {children}
    </button>
  );
}
