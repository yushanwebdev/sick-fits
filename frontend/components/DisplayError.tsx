import styled from 'styled-components';
import React from 'react';
import { ApolloError } from '@apollo/client';

const ErrorStyles = styled.div`
  padding: 2rem;
  background: white;
  margin: 2rem 0;
  border: 1px solid rgba(0, 0, 0, 0.05);
  border-left: 5px solid red;
  p {
    margin: 0;
    font-weight: 100;
  }
  strong {
    margin-right: 1rem;
  }
`;

const DisplayError = ({ error }: { error: ApolloError }) => {
  if (!error || !error.message) return null;

  return (
    <ErrorStyles>
      <p data-test="graphql-error">
        <strong>Shoot!</strong>
        {error.message.replace('GraphQL error: ', '')}
      </p>
    </ErrorStyles>
  );
};

export default DisplayError;
