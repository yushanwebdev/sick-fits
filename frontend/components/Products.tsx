import { gql, useQuery } from '@apollo/client';
import styled from 'styled-components';
import Product from './Product';

export const ALL_PRODUCTS_QUERY = gql`
  query GetAllProducts {
    allProducts {
      id
      name
      price
      description
      photo {
        id
        image {
          publicUrlTransformed
        }
      }
    }
  }
`;

const ProductListStyles = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-gap: 60px;
`;

export default function Products() {
  // TODO: fix this error
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
  const { data, loading, error } = useQuery(ALL_PRODUCTS_QUERY);

  if (loading) {
    return <p>Loading...</p>;
  }

  if (error) {
    return <p>Error: {error.message}</p>;
  }

  return (
    <ProductListStyles>
      {data?.allProducts.map((product) => (
        <Product key={product.id} product={product} />
      ))}
    </ProductListStyles>
  );
}
