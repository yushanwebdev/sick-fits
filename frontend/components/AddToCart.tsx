/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable react/prop-types */
import { gql, useMutation } from '@apollo/client';
import { CURRENT_USER_QUERY } from '../lib/useUser';

const ADD_TO_CART_MUTATION = gql`
  mutation AddToCart($id: ID!) {
    addToCart(productId: $id) {
      id
      quantity
    }
  }
`;

export default function AddToCart({ id }) {
  const [addToCart, { loading }] = useMutation(ADD_TO_CART_MUTATION, {
    variables: {
      id,
    },
    refetchQueries: [CURRENT_USER_QUERY],
  });

  return (
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    <button type="button" onClick={addToCart} disabled={loading}>
      Add{loading && 'ing'} to Cart ðŸ›’
    </button>
  );
}
