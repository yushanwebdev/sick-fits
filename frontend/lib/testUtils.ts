import casual from "casual";
import { PAGINATION_QUERY } from "../components/Pagination";

// seed it so we get consistent results
casual.seed(777);

const fakeItem = () => ({
  __typename: "Item",
  id: "abc123",
  price: 5000,
  user: null,
  photo: {
    __typename: "Photo",
    id: "abc123",
    altText: "dogs are best",
    image: {
      __typename: "Image",
      publicUrlTransformed: "dog.jpg",
    },
  },
  name: "dogs are best",
  description: "dogs",
});

const fakeUser = (overrides?: Object) => ({
  __typename: "User",
  id: "4234",
  name: casual.name,
  email: casual.email,
  permissions: ["ADMIN"],
  orders: [],
  cart: [],
  role: {
    canManageCart: true,
    canManageUsers: true,
    canManageProducts: true,
  },
  ...overrides,
});

const fakeOrderItem = () => ({
  __typename: "OrderItem",
  id: casual.uuid,
  image: {
    image: `${casual.word}.jpg`,
  },
  name: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words(),
});

const fakeOrder = () => ({
  __typename: "Order",
  id: "ord123",
  charge: "ch_123",
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: "2022-12-11T20:16:13.797Z",
  user: fakeUser(),
});

const fakeCartItem = (overrides?: Object) => ({
  __typename: "CartItem",
  id: "omg123",
  quantity: 3,
  product: fakeItem(),
  user: fakeUser(),
  ...overrides,
});

function makePaginationMocksFor(length) {
  return [
    {
      request: { query: PAGINATION_QUERY },
      result: {
        data: {
          _allProductsMeta: {
            count: length,
          },
          itemsConnection: {
            __typename: "aggregate",
            aggregate: {
              count: length,
              __typename: "count",
            },
          },
        },
      },
    },
  ];
}

export {
  makePaginationMocksFor,
  fakeItem,
  fakeUser,
  fakeCartItem,
  fakeOrder,
  fakeOrderItem,
};
