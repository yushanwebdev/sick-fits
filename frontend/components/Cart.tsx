/* eslint-disable @typescript-eslint/no-unsafe-argument */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import styled from 'styled-components';
import calcTotalPrice from '../lib/calcTotalPrice';
import formatMoney from '../lib/formatMoney';
import { useUser } from '../lib/useUser';
import CartStyles from './styles/CartStyles';
import Supreme from './styles/Supreme';

interface CartItemProps {
  cartItem: {
    quantity: number;
    id: string;
    product: {
      name: string;
      price: number;
      photo: {
        image: {
          publicUrlTransformed: string;
        };
      };
    };
  };
}

const CartItemStyles = styled.li`
  padding: 1rem 0;
  border-bottom: 1px solid ${({ theme }) => theme.colors.lightGray};
  display: grid;
  grid-template-columns: auto 1fr auto;
  img {
    margin-right: 1rem;
  }
  h3,
  p {
    margin: 0;
  }
`;

function CartItem({ cartItem }: CartItemProps) {
  const { product } = cartItem;

  return (
    <CartItemStyles>
      <img
        src={product.photo.image.publicUrlTransformed}
        alt={product.name}
        width="100"
      />
      <div>
        <h3>{product.name}</h3>
        <p>
          {formatMoney(product.price * cartItem.quantity)} -{' '}
          <em>
            {cartItem.quantity} &times; {formatMoney(product.price)} each
          </em>
        </p>
      </div>
    </CartItemStyles>
  );
}

export default function Cart() {
  const me = useUser();

  if (!me) return null;

  console.log('me', me);

  return (
    <CartStyles open>
      <header>
        <Supreme>{me.name}'s Cart</Supreme>
      </header>
      <ul>
        {me.cart.map((cartItem) => (
          <CartItem key={cartItem.id} cartItem={cartItem} />
        ))}
      </ul>
      <footer>
        <p>{formatMoney(calcTotalPrice(me.cart))}</p>
      </footer>
    </CartStyles>
  );
}
