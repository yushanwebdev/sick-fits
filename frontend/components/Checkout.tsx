import { gql, useMutation } from "@apollo/client";
import {
  CardElement,
  Elements,
  useElements,
  useStripe,
} from "@stripe/react-stripe-js";
import { loadStripe, StripeError } from "@stripe/stripe-js";
import { useRouter } from "next/router";
import nProgress from "nprogress";
import { useState } from "react";
import styled from "styled-components";
import { useCart } from "../lib/cartState";
import { CURRENT_USER_QUERY } from "../lib/useUser";
import SickButton from "./styles/SickButton";

const CheckoutFormStyles = styled.form`
  box-shadow: 0 1px 2px 2px rgba(0, 0, 0, 0.04);
  border: 1px solid rgba(0, 0, 0, 0.06);
  border-radius: 5px;
  padding: 1rem;
  display: grid;
  grid-gap: 1rem;
`;

const CREATE_ORDER_MUTATION = gql`
  mutation CreateOrderMutation($token: String!) {
    checkout(token: $token) {
      id
      charge
      total
      items {
        id
        name
      }
    }
  }
`;

const stripeLib = loadStripe(process.env.NEXT_PUBLIC_STRIPE_KEY);

function CheckoutForm() {
  const [error, setError] = useState<StripeError | undefined>();
  const [loading, setLoading] = useState(false);
  const stripe = useStripe();
  const elements = useElements();
  const [checkout, { error: graphQLError }] = useMutation(
    CREATE_ORDER_MUTATION,
    {
      refetchQueries: [
        {
          query: CURRENT_USER_QUERY,
        },
      ],
    }
  );
  const router = useRouter();
  const { closeCart } = useCart();

  async function handleSubmit(e) {
    // 1. Stop the form from submitting & turn the loader on
    e.preventDefault();
    setLoading(true);
    console.log("We gotta to do some work...");
    // 2. Start the page transition
    nProgress.start();
    // 3. Create the payment method via stripe (Token comes back here if successful)
    const { error: stripeError, paymentMethod } =
      await stripe.createPaymentMethod({
        type: "card",
        card: elements.getElement(CardElement),
      });

    // 4. Handle any errors from stripe
    if (stripeError) {
      setError(stripeError);
      nProgress.done();
      return; // stops the checkout from happening
    }
    // 5. Send the token from step 3 to our keystone server, via a custom mutation
    console.log("paymentMethod", paymentMethod);
    const order = await checkout({
      variables: {
        token: paymentMethod.id,
      },
    });
    console.log("Finished with the order!!", order);
    // 6. Change the page to view the order
    router.push({
      pathname: "/order/[id]",
      query: {
        id: order.data.checkout.id,
      },
    });
    // 7. Close the cart
    closeCart();

    // 8. turn the loader off
    setLoading(false);
    nProgress.done();
  }

  return (
    <CheckoutFormStyles onSubmit={handleSubmit}>
      {error && (
        <p
          style={{
            fontSize: "1.2rem",
          }}
        >
          {error.message}
        </p>
      )}
      {graphQLError && (
        <p
          style={{
            fontSize: "1.2rem",
          }}
        >
          {graphQLError.message}
        </p>
      )}
      <CardElement />
      <SickButton>Check Out Now</SickButton>
    </CheckoutFormStyles>
  );
}

export default function Checkout() {
  return (
    <Elements stripe={stripeLib}>
      <CheckoutForm />
    </Elements>
  );
}
