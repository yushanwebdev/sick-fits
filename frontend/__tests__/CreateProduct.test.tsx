import { MockedProvider } from "@apollo/client/testing";
import { render, screen } from "@testing-library/react";
import CreateProduct, {
  CREATE_PRODUCT_MUTATION,
} from "../components/CreateProduct";
import userEvent from "@testing-library/user-event";
import { fakeItem } from "../lib/testUtils";
import { ALL_PRODUCTS_QUERY } from "../components/Products";
import mockRouter from "next-router-mock";

jest.mock("next/router", () => require("next-router-mock"));

const item = fakeItem();

describe("<CreateProduct />", () => {
  it("renders and matches snapshot", () => {
    const { container, debug } = render(
      <MockedProvider>
        <CreateProduct />
      </MockedProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it("handles the updating", async () => {
    // 1. render the form out
    const { container, debug } = render(
      <MockedProvider>
        <CreateProduct />
      </MockedProvider>
    );
    // 2. type into the boxes
    await userEvent.type(screen.getByPlaceholderText(/Name/i), item.name);
    await userEvent.type(
      screen.getByPlaceholderText(/Price/i),
      item.price.toString()
    );
    await userEvent.type(
      screen.getByPlaceholderText(/Description/i),
      item.description
    );

    // 3. check that those boxes are populated!
    expect(screen.getByDisplayValue(item.name)).toBeInTheDocument();
    expect(screen.getByDisplayValue(item.price)).toBeInTheDocument();
    expect(screen.getByDisplayValue(item.description)).toBeInTheDocument();
  });

  it("creates the items when the form is submitted", async () => {
    // create the mocks for this one
    const mocks = [
      {
        request: {
          query: CREATE_PRODUCT_MUTATION,
          variables: {
            name: item.name,
            description: item.description,
            image: "",
            price: item.price,
          },
        },
        result: {
          data: {
            createProduct: {
              ...item, // add fake item fields
              id: "abc123",
              __typename: "Item",
            },
          },
        },
      },
      {
        request: {
          query: ALL_PRODUCTS_QUERY,
          variables: {
            skip: 0,
            first: 2,
          },
        },
        result: {
          data: {
            allProducts: [item],
          },
        },
      },
    ];

    const { container, debug } = render(
      <MockedProvider mocks={mocks}>
        <CreateProduct />
      </MockedProvider>
    );

    // Type into the inputs
    await userEvent.type(screen.getByPlaceholderText(/Name/i), item.name);
    await userEvent.type(
      screen.getByPlaceholderText(/Price/i),
      item.price.toString()
    );
    await userEvent.type(
      screen.getByPlaceholderText(/Description/i),
      item.description
    );

    // Submit it and see if the page change has been called
    await userEvent.click(screen.getByText(/Add Product/));

    expect(mockRouter).toMatchObject({
      asPath: `/product/abc123`,
    });
  });
});
