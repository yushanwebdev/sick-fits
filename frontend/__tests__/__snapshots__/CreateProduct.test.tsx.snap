// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateProduct /> renders and matches snapshot 1`] = `
<div>
  <form
    class="sc-23a63c62-0 bqKLMU"
  >
    <fieldset
      aria-busy="false"
    >
      <label>
        Image
        <input
          id="image"
          name="image"
          type="file"
        />
      </label>
      <label>
        Name
        <input
          id="name"
          name="name"
          placeholder="Name"
          required=""
          type="text"
          value=""
        />
      </label>
      <label>
        Price
        <input
          id="price"
          name="price"
          placeholder="Price"
          required=""
          type="number"
          value="0"
        />
      </label>
      <label>
        Description
        <textarea
          id="description"
          name="description"
          placeholder="Description"
          required=""
        />
      </label>
      <button
        type="submit"
      >
        + Add Product
      </button>
    </fieldset>
  </form>
</div>
`;
